Add UTxO snapshot API endpoint and CLI command.

jonathanknowles opened the pull request:
# Issue Number

Related to ADP-890.

# Overview

This PR adds the `getWalletUtxoSnapshot` API endpoint and associated CLI command.

The intended use is for debugging and development.

The endpoint generates a complete snapshot of a wallet's UTxO set, as an unordered list of UTxO entries.

It produces JSON output similar to the following:
```json
"entries": [
  { "ada":         {"quantity": 8, "unit": "lovelace"}
  , "ada_minimum": {"quantity": 1, "unit": "lovelace"}
  , "assets": []
  },
  { "ada":         {"quantity": 4, "unit": "lovelace"}
  , "ada_minimum": {"quantity": 2, "unit": "lovelace"}
  , "assets": [
      { "policy_id": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
      , "asset_name": "APPLE"
      , "quantity": 10
      }
    ]
  },
  { "ada":         {"quantity": 4, "unit": "lovelace"}
  , "ada_minimum": {"quantity": 2, "unit": "lovelace"}
  , "assets": [
      { "policy_id": "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"
      , "asset_name": "BANANA"
      , "quantity": 20
      }
    ]
  }
]
```

jonathanknowles committed:
Add `GetWalletUtxoSnapshot` API endpoint.

jonathanknowles committed:
Add `GetWalletUtxoSnapshot` CLI command.

jonathanknowles committed:
Add golden data for `ApiWalletUtxoSnapshot`.

jonathanknowles committed:
Add API specification for `getWalletUtxoSnapshot`.

Commit:
Add integration tests for the API

Commit:
Add integration test for CLI

piotr-iohk reviewed (approved):
lgtm

paweljakubas reviewed (approved):
lgtm!

I would suggest more integration test(s) - can be in some future PR:
1. setup fixture wallet and verify snapshot. Then realize tx and when, both pending/in_ledger, verify utxo snapshots for both source and target wallet 
2. Do some funds flow testing and compare expectations of utxo snapshot and utxo statistics
3. Do something interesting with non-ada things and check like in point 1. And like point 2 if we have already utxo statistics for native assets

jonathanknowles committed:
Use consistent names for API integration tests.

This change adjusts the wallet UTxO snapshot API integration tests to use
the common match string: `WALLET_UTXO_SNAPSHOT`.

This is consistent with the API function name.

We also extract out lists of wallets into separate definitions, to avoid
excessively long lines.

jonathanknowles committed:
Use consistent name for `Api.Link` function.

The name `getWalletUtxoSnapshot` is consistent with the name of the API
function.

jonathanknowles committed:
Fix formatting in `getWalletUtxoSnapshotViaCLI`.

jonathanknowles committed:
Use consistent names for CLI integration tests.

Similarly to the API tests, we use the `WALLET_UTXO_SNAPSHOT` label.

rvl commented on lib/core/src/Cardano/Wallet/Api/Link.hs:
```diff
@@ -237,6 +238,21 @@ getUTxOsStatistics w = discriminate @style
+getWalletUtxoSnapshot w = discriminate @style
+    (endpoint @Api.GetWalletUtxoSnapshot (wid &))
+    (endpoint @Api.GetByronWalletUtxoSnapshot (wid &))
+    (notSupported "Shared")
```
Why not supported for shared wallets? They are have UTxO, once active.

jonathanknowles replied:
> Why not supported for shared wallets? They are have UTxO, once active.

Good point.

@piotr-iohk, do you know of a good reason to not support shared wallets?

piotr-iohk replied:
Nope, my guess was that we will support them eventually at some point (once they are active). :)
If it is possible to wire them up right now then, yes, why not.

rvl commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -697,6 +699,20 @@ newtype ApiWalletPassphrase = ApiWalletPassphrase
     } deriving (Eq, Generic, Show)
       deriving anyclass NFData
 
+newtype ApiWalletUtxoSnapshot = ApiWalletUtxoSnapshot
```
I think it's possible to avoid the extra type by just using `ApiT [ApiWalletUtxoSnapshotEntry]`.

rvl commented on specifications/api/swagger.yaml:
```diff
@@ -2255,6 +2255,39 @@ components:
+      required:
+        - entries
+      properties:
+        entries:
```
The extra level of wrapping in an object isn't necessary, is it?

jonathanknowles replied:
> The extra level of wrapping in an object isn't necessary, is it?

I think this is arguable either way.

I agree with you that using a list would be enough to capture the current use case.

But on the other hand, using an object gives us the flexibility to add extra fields relating to a snapshot at a later time, without having to break compatibility with clients of the API (including ourselves).

From an API aesthetics point of view, I feel it's clearer to be able to point to a single type and say "this single type represents a complete UTxO snapshot", similarly to how we use `UTxO` to represent a complete UTxO set.

rvl commented on specifications/api/swagger.yaml:
```diff
@@ -4437,6 +4479,19 @@ paths:
+      tags: ["Wallets"]
+      summary: A snapshot of the wallet's UTxO set
+      description: |
+        Generate a snapshot of the wallet's UTxO set.
```
Most endpoints have a `status: stable` note.
Since this is intended to be used for debugging, better to not imply stability - perhaps add something like `status: unstable`.

jonathanknowles replied:
> Most endpoints have a status: stable note.
> Since this is intended to be used for debugging, better to not imply stability - perhaps add
> something like status: unstable.

Good point!

rvl reviewed (commented):
Nice addition, thanks :+1:
